### PR TITLE
#2382 - Redis recovery command

### PR DESCRIPTION
**Pull request by guru-aot**
Created a make command from the below recommendation from the Redis channel in rocket chat and it seems to recover our Redis without deleting it completely in the environment and creating them again.

image

Note: This command had to be run every time, when we put down the Redis manually or any disruptions that restart the Redis pods. Parallel to this, if we have a Sysdig alert which checks the health of the Redis pods in the environment, would help us in any delays to the application queues processing.

Also wiki has been appropriately updated
https://github.com/bcgov/SIMS/wiki/DevOps-and-Running-the-Application#redis-cluster-failing-to-recover

### CODE REVIEW

File: devops/Makefile [hunk 1]
```diff
@@ -291,6 +291,12 @@ init-redis-cluster:
 	REDIS_NODES=$$(oc get pods -n $(NAMESPACE) -l app=redis -o jsonpath='{range .items[*]}{.status.podIP}:6379 ' | sed 's/ :6379 $$/ /'); \
 	oc exec -n $(NAMESPACE) -it redis-0 -- redis-cli -a $$REDIS_PASSWORD --cluster create $$REDIS_NODES --cluster-replicas 1
 
+# Command to tell one Redis node to connect to another node and join the cluster. 
+# This is typically done when you are setting up a Redis Cluster or adding new nodes to an existing cluster to ensure they are aware of each other and can communicate effectively.
+redis-cluster-meet:
+	REDIS_PASSWORD=$$(oc get secret -n $(NAMESPACE) redis-creds -o jsonpath='{.data.password}' | base64 -d); \
+  oc get pods -n $(NAMESPACE) -l app=redis -o jsonpath='{range.items[*]}{.status.podIP} 6379 {end}' | xargs -n2 oc exec -n $(NAMESPACE) -it redis-0 -- redis-cli -a $$REDIS_PASSWORD -c CLUSTER MEET
```
Comment (cditcher) [inline devops/Makefile:298]: Should there be a colon before the port number? jsonpath='{range.items[*]}{.status.podIP}:6379

**Review by cditcher — COMMENTED** [review]
Should there be a colon

File: devops/Makefile (hunk 1 above)
Comment (guru-aot) [reply devops/Makefile:298]: I guess the command takes the variables as IP 6379, in the previous cluster create command we used ip:6379 as the command requires that. So i believe in this command this is not required. If we supply with the : , we get the below error.
ERR Invalid TCP base port specified: <ip>:6379

**Review by dheepak-aot — APPROVED** [review]
Looks good. Great work 👍

**Review by ann-aot — APPROVED** [review]
👍

File: devops/Makefile [hunk 2]
```diff
@@ -291,6 +291,12 @@ init-redis-cluster:
 	REDIS_NODES=$$(oc get pods -n $(NAMESPACE) -l app=redis -o jsonpath='{range .items[*]}{.status.podIP}:6379 ' | sed 's/ :6379 $$/ /'); \
 	oc exec -n $(NAMESPACE) -it redis-0 -- redis-cli -a $$REDIS_PASSWORD --cluster create $$REDIS_NODES --cluster-replicas 1
 
+# Command to tell one Redis node to connect to another node and join the cluster. 
+# This is typically done when you are setting up a Redis Cluster or adding new nodes to an existing cluster to ensure they are aware of each other and can communicate effectively.
+redis-cluster-meet:
```
Comment (andrewsignori-aot) [inline devops/Makefile:296]: Thanks for creating the make command. I would recommend that we also create a GitHub Action to allow easy execution.

File: devops/Makefile [hunk 3]
```diff
@@ -291,6 +291,12 @@ init-redis-cluster:
 	REDIS_NODES=$$(oc get pods -n $(NAMESPACE) -l app=redis -o jsonpath='{range .items[*]}{.status.podIP}:6379 ' | sed 's/ :6379 $$/ /'); \
 	oc exec -n $(NAMESPACE) -it redis-0 -- redis-cli -a $$REDIS_PASSWORD --cluster create $$REDIS_NODES --cluster-replicas 1
 
+# Command to tell one Redis node to connect to another node and join the cluster. 
+# This is typically done when you are setting up a Redis Cluster or adding new nodes to an existing cluster to ensure they are aware of each other and can communicate effectively.
```
Comment (andrewsignori-aot) [inline devops/Makefile:295]: Is there any harm in executing the command at the wrong moment?
For instance, if Redis is not working properly, is it ok to execute the command as a first attempt to recover the Cluster?

File: devops/Makefile (hunk 1 above)
Comment (andrewsignori-aot) [inline devops/Makefile:298]: Not a problem for this PR but port 6379 is not centrilized anywhere, I believe.

File: devops/Makefile (hunk 3 above)
Comment (guru-aot) [reply devops/Makefile:295]: No i have tried it, it doesnt do anything if everything is working

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work and research. Just a few clarifications and please lets have a demo to see Redis being recovered using this approach.

File: devops/Makefile (hunk 3 above)
Comment (andrewsignori-aot) [reply devops/Makefile:295]: Good to know 😉

**Review by andrepestana-aot — APPROVED** [review]
Looks good!

File: .github/workflows/redis-recovery.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  recoverRedis:
+    name: Redis recovery in Openshift
+    runs-on: ubuntu-latest
+    environment: ${{ inputs.environment }}
+    env:
+      NAMESPACE: ${{ secrets.OPENSHIFT_ENV_NAMESPACE }}
+    steps:
+      - name: Print env
+        run: |
+          echo NAMESPACE: $NAMESPACE
```
Comment (andrewsignori-aot) [inline .github/workflows/redis-recovery.yml:26]: GitHub will not print a value that came from a secret so there is no point in printing it, IMO.

File: .github/workflows/redis-recovery.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      - name: Checkout Target Branch
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ inputs.gitRef }}
+      - name: Log in to OpenShift
+        run: |
+          oc login --token=${{ secrets.SA_TOKEN }} --server=https://api.silver.devops.gov.bc.ca:6443
```
Comment (andrewsignori-aot) [inline .github/workflows/redis-recovery.yml:33]: Please use the below instead.
```
oc login --token=${{ secrets.SA_TOKEN }} --server=${{ vars.OPENSHIFT_CLUSTER_URL }}
```

File: .github/workflows/redis-recovery.yml
Comment (andrewsignori-aot) [inline .github/workflows/redis-recovery.yml:1]: Every GitHub action is "categorized" as "env-setup" or "release" which adds a prefix to the file name and to the name itself.
Can we rename this file to `env-setup-redis-recovery.yml` and the name to `Env Setup - Redis recovery in Openshift`

**Review by andrewsignori-aot — APPROVED** [review]
Looks great, thanks again for adding the GitHub action, looks good 👍

**Pull request merged**